src/Compilers/CSharp/Test/Semantic/Semantics/NameOfTests.cs

```diff
@@ -954,6 +956,315 @@ public static void Main(string[] args)
             Assert.Equal(0, symbolInfo.CandidateSymbols.Length);
         }
 
+        [Fact]
+        public void SymbolInfo_InstanceMemberFromStatic_Flat()
+        {
+            var source = """
+                public class C
+                {
+                    public int Property { get; }
+                    public int Field;
+                    public event System.Action Event;
+
+                    public static string StaticField =
+                        nameof(Property) +
+                        nameof(Field) +
+                        nameof(Event);
+                }
+                """;
+            var comp = CreateCompilation(source).VerifyDiagnostics();
+
+            var cProperty = comp.GetMember("C.Property");
+            var cField = comp.GetMember("C.Field");
+            var cEvent = comp.GetMember("C.Event");
+
+            var tree = comp.SyntaxTrees.Single();
+            var tree2 = SyntaxFactory.ParseSyntaxTree(source + " ");
+
+            var initializer = tree2.GetRoot().DescendantNodes().OfType<EqualsValueClauseSyntax>().Single();
+
+            var nameofCalls = getNameOfCalls(tree);
+            Assert.Equal(3, nameofCalls.Length);
+            var nameofCalls2 = getNameOfCalls(tree2);
+            Assert.Equal(3, nameofCalls2.Length);
+
+            var model = comp.GetSemanticModel(tree);
+
+            verify(0, "Property", cProperty);
+            verify(1, "Field", cField);
+            verify(2, "Event", cEvent);
+
+            void verify(int index, string expression, Symbol symbol)
+            {
+                var argument = nameofCalls[index].ArgumentList.Arguments.Single().Expression;
+                Assert.Equal(expression, argument.ToString());
+
+                verifySymbolInfo(model.GetSymbolInfo(argument));
+
+                var argument2 = nameofCalls2[index].ArgumentList.Arguments.Single().Expression;
+                Assert.Equal(expression, argument2.ToString());
+
+                Assert.True(model.TryGetSpeculativeSemanticModel(initializer.Position, initializer, out var model2));
+
+                verifySymbolInfo(model2.GetSymbolInfo(argument2));
+
+                verifySymbolInfo(model.GetSpeculativeSymbolInfo(argument2.Position, argument2, SpeculativeBindingOption.BindAsExpression));
+
+                Assert.True(model.GetSpeculativeSymbolInfo(argument2.Position, argument2, SpeculativeBindingOption.BindAsTypeOrNamespace).IsEmpty);
+
+                void verifySymbolInfo(SymbolInfo symbolInfo)
```

Comment:
nit: when possible, let's avoid nested local functions (for future consideration)